### ISSUE TITLE
Cannot compile with Visual Studio 2017 RC (solved)

### ISSUE THREAD

**Issue by bennyyip** [Valloric/ycmd#656]
I tried to compile YouCompleteMe using the `install.py` script, but CMake cannot find the MSVC compiler. I was using CMake 3.7, the lastest CMake release. It seems that it cannot detect the new compiler. 
```powershell
λ python .\install.py --clang-completer
Searching Python 3.5 libraries...
Found Python library: C:\Users\yeben\Anaconda3\libs\python35.lib
Found Python headers folder: C:\Users\yeben\Anaconda3\include
-- The C compiler identification is unknown
-- The CXX compiler identification is unknown
CMake Error at CMakeLists.txt:26 (project):
  No CMAKE_C_COMPILER could be found.



CMake Error at CMakeLists.txt:26 (project):
  No CMAKE_CXX_COMPILER could be found.



-- Configuring incomplete, errors occurred!
See also "C:/Users/yeben/AppData/Local/Temp/ycm_build_4gbmd4_v/CMakeFiles/CMakeOutput.log".
See also "C:/Users/yeben/AppData/Local/Temp/ycm_build_4gbmd4_v/CMakeFiles/CMakeError.log".
ERROR: the build failed.

NOTE: it is *highly* unlikely that this is a bug but rather
that this is a problem with the configuration of your system
or a missing dependency. Please carefully read CONTRIBUTING.md
and if you're sure that it is a bug, please raise an issue on the
issue tracker, including the entire output of this script
and the invocation line used to run it.
```
Then I use the CMake installed with VS 2017 instead, and modify the `ycmd/build.py`. 
```Python
# ycmd/build.py
def BuildYcmdLib( args ):
  build_dir = mkdtemp( prefix = 'ycm_build_' )

  try:
    # full_cmake_args = [ '-G', GetGenerator( args ) ]
    full_cmake_args = [ '-G', "Visual Studio 15 2017 Win64" ]  # hack
    full_cmake_args.extend( CustomPythonCmakeArgs() )
    full_cmake_args.extend( GetCmakeArgs( args ) )
    full_cmake_args.append( p.join( DIR_OF_THIS_SCRIPT, 'cpp' ) )
```

I re-compiled, CMake successfully detected MSVC, but Boost failed to complie. Because it considers MSVC 1910 as an unknown version.
```C++
// ycmd\cpp\BoostParts\boost\config\compiler\visualc.hpp
//
// last known and checked version is 19.00.23026 (VC++ 2015 RTM):
#if (_MSC_VER > 1900)
#  if defined(BOOST_ASSERT_CONFIG)
#     error "Unknown compiler version - please run the configure tests and report the results"
#  else
#     pragma message("Unknown compiler version - please run the configure tests and report the results")
#  endif
#endif
```
I changed 1900 to 1910 and finally solved it.

I don't know how actually the build script work and I just found a hacky way to solve it. I wish you guys update the build script, which can be used to compile ycmd on windows easier :)

Not English native speaker, sorry for bad grammar.

**Comment by vheon** [comment]
>     full_cmake_args = [ '-G', "Visual Studio 15 2017 Win64" ]  # hack

To do this the proper way to change the `--msvc` flag to accept `15` and `GetGenerator` to accept that and return `"Visual Studio 15"` the `"Win64"` will be inserted by inferring the machine architecture.

> Because it considers MSVC 1910 as an unknown version.
> ...
> I changed 1900 to 1910 and finally solved it.

Unfortunately that is actually Boost, we only download it, so this has to be reported and fixed upstream. Does a new release of boost exists?

**Comment by micbou** [comment]
I suggest to wait for the final release of Visual Studio 2017 before supporting it in the `build.py` script. Concerning Boost, this will probably be fixed in its next release (1.63.0), which is planned next month.

**Comment by BenjaminHao** [comment]
GREAT! Just installed VS2017 and this is exactly the solution for my problem.
Thx!
